sw-emulator/lib/periph/src/dma/encryption_engine.rs

```diff
@@ -0,0 +1,820 @@
+
+Abstract:
+
+    File contains Encryption Engine Implementation
+
+--*/
+
+use caliptra_emu_bus::{BusError, ReadWriteRegister};
+use caliptra_emu_derive::Bus;
+use caliptra_emu_types::{RvData, RvSize};
+use tock_registers::interfaces::Readable;
+use tock_registers::register_bitfields;
+use smlang::statemachine;
+use std::collections::HashMap;
+
+#[cfg(test)]
+macro_rules! assert_ctrl_eq {
+    ($encryption_engine:ident, $target:expr) => {
+        assert_eq!($encryption_engine.read(RvSize::Word, CTRL).unwrap(), $target.into());
+    }
+}
+
+register_bitfields! [
+    u32,
+    Control [
+        EXE OFFSET(0) NUMBITS(1) [
+            RUN = 1,
+        ],
+        DONE OFFSET(1) NUMBITS(1) [
+            DONE = 1,
+        ],
+        CMD OFFSET(2) NUMBITS(4) [
+            LOAD_MEK = 1,
+            UNLOAD_MEK = 2,
+            ZEROIZE = 3,
+        ],
+        ERR OFFSET(16) NUMBITS(4) [
+            NO_ERROR = 0,
+            INVALID_COMMAND = 1,
+            VENDOR_SPECIFIC_KEY_CACHE_FULL = 4,
+            VENDOR_SPECIFIC_MEK_NOT_FOUND = 5,
+            VENDOR_SPECIFIC_REDUNDANT_METADATA = 6,
+        ],
+        RDY OFFSET(31) NUMBITS(1) [
+            NOT_READY = 0,
+            READY = 1,
+        ]
+    ],
+];
+
+impl TryFrom<u32> for Control::ERR::Value {
+    type Error = ();
+
+    fn try_from(value: u32) -> Result<Self, ()> {
+        match value {
+            0x0 => Ok(Control::ERR::Value::NO_ERROR),
+            0x1 => Ok(Control::ERR::Value::INVALID_COMMAND),
+            0x4 => Ok(Control::ERR::Value::VENDOR_SPECIFIC_KEY_CACHE_FULL),
+            0x5 => Ok(Control::ERR::Value::VENDOR_SPECIFIC_MEK_NOT_FOUND),
+            0x6 => Ok(Control::ERR::Value::VENDOR_SPECIFIC_REDUNDANT_METADATA),
+            _ => Err(()),
+        }
+    }
+}
+
+impl TryFrom<u32> for Control::CMD::Value {
+    type Error = ();
+
+    fn try_from(value: u32) -> Result<Self, ()> {
+        match value {
+            0x1 => Ok(Control::CMD::Value::LOAD_MEK),
+            0x2 => Ok(Control::CMD::Value::UNLOAD_MEK),
+            0x3 => Ok(Control::CMD::Value::ZEROIZE),
+            _ => Err(()),
+        }
+    }
+}
+
+statemachine! {
+    transitions: {
+        *Idle + Initalize / initalize_encryption_engine = Initializing,
+        Idle + InitFail / handle_fatal_error = Fatal,
+        Initializing + InitSuccess = Ready,
+        Ready + WriteCommandExecution / handle_execute = Processing,
+        Ready + ProcessFail / handle_fatal_error = Fatal,
+        Processing + ProcessSuccess / execution_done = WaitClear,
+        WaitClear + WriteDone / handle_clear = Clearing,
+        WaitClear + ClearFail / handle_fatal_error = Fatal,
+        Clearing + ClearSuccess / clear_done = Ready,
+    }
+}
+
+// Keep small key cache size for test
+const KEY_CACHE_SIZE: usize = 8;
+
+#[derive(Debug)]
+struct Context {
+    key_cache: HashMap<[u32; 5], ([u32; 8], [u32; 16])>,
+    mek: [u32; 16],
+    metadata: [u32; 5],
+    auxiliary_data: [u32; 8],

```

Comment:
I think it would be useful to wrap these in a new type for readability

For example:

```
type Mek = [u8; 16];
type Metadata = [u8; 5];
type Auxdata = [u8; 8];

key_cache: HashMap<Metadata, (AuxData, Mek)>
```